FRAuthenticator/FRAuthenticator/Push/FRAPushHandler.swift

```diff
@@ -68,35 +68,41 @@ public class FRAPushHandler: NSObject {
-            
-            if let mechanism = FRAClient.storage.getMechanismForUUID(uuid: notification.mechanismUUID) {
-                if try FRCompactJWT.verify(jwt: jwt, secret: mechanism.secret) == false {
-                    FRALog.e("Failed to verify given JWT in remote-notification payload; returning nil")
+            // Check if notification with given messageId already exists
+            if let notification = FRAClient.storage.getNotificationByMessageId(messageId: messageId) {
+                FRALog.v("Received remote-notification with messageId: \(messageId) already exists in StorageClient; returning the existing PushNotification object")
+                return notification
+            } else {
+                // Extract JWT payload
+                FRALog.v("Starts extracting JWT payload: \(jwt)")
+                let jwtPayload = try FRCompactJWT.extractPayload(jwt: jwt)
+                FRALog.v("JWT payload is extracted: \(jwtPayload)")
+                
+                // Construct and save Notification object
+                FRALog.v("PushNotification object created - messageId:\(messageId), payload: \(jwtPayload)")
+                let notification = try PushNotification(messageId: messageId, payload: jwtPayload)
+                
+                if let mechanism = FRAClient.storage.getMechanismForUUID(uuid: notification.mechanismUUID) {
+                    if try FRCompactJWT.verify(jwt: jwt, secret: mechanism.secret) == false {
+                        FRALog.e("Failed to verify given JWT in remote-notification payload; returning nil")
+                        return nil
+                    }
+                    FRALog.v("Verification of JWT in remote-notification payload with PushMechanism's secret")
+                }
+                else {
+                    FRALog.e("Failed to retrieve PushMechanism object from StorageClient; returning null")
                     return nil
                 }
-                FRALog.v("Verification of JWT in remote-notification payload with PushMechanism's secret")
-            }
-            else {
-                FRALog.e("Failed to retrieve PushMechanism object from StorageClient; returning null")
-                return nil
-            }
-            
-            if FRAClient.storage.setNotification(notification: notification) {
-                FRALog.v("PushNotification object is created and saved into StorageClient")
-            }
-            else {
-                FRALog.w("PushNotification object failed to be stored into StorageClient")
+                
+                if FRAClient.storage.setNotification(notification: notification) {
+                    FRALog.v("PushNotification object is created and saved into StorageClient")
+                }
+                else {
+                    FRALog.w("PushNotification object failed to be stored into StorageClient")
```

Comment:
Should we not treat that as an error and return nil? If we end up on this situation, is the Push actionable or would the app be stuck?

Reply:
good catch! we should return nil here. It's unlikely we will have an exception while persisting it because the default storage is the Keychain. However, if a developer decide to use SQLite or any other storage, it may happen, so we need to confirm